client/memory.cpp

```diff
@@ -84,6 +85,46 @@ clEnqueueFillBuffer(cl_command_queue command_queue, cl_mem buffer,
 	}
 }
 
+SO_EXPORT CL_API_ENTRY void* CL_API_CALL
+clEnqueueMapBuffer(cl_command_queue command_queue, cl_mem buffer, cl_bool blocking_map,
+                   cl_map_flags map_flags, size_t offset, size_t size,
+                   cl_uint num_events_in_wait_list, const cl_event* event_wait_list,
+                   cl_event* event, cl_int *errcode_ret) CL_API_SUFFIX__VERSION_1_0
+{
+	if (buffer == nullptr) {
+		*errcode_ret = CL_INVALID_MEM_OBJECT;
+		return nullptr;
+	}
+
+	auto conn = gConnection.get();
+	auto &buf = conn.registerBuffer(GetID(buffer));
+	void *ptr;
+
+	try {
+		buf.data.reset(new uint8_t[size]);
+		buf.flags = map_flags;
+		buf.offset = offset;
+		buf.size = size;
+		ptr = reinterpret_cast<void*>(buf.data.get());
+	} catch (...) {
+		*errcode_ret = CL_OUT_OF_HOST_MEMORY;
+		return nullptr;
+	}
+
+	cl_int retain = clRetainMemObject(buffer);
+	if (retain != CL_SUCCESS) {
+		*errcode_ret = CL_INVALID_MEM_OBJECT;
+		return nullptr;
+	}
```

Comment:
Should the mem object be retained after a possible clEnqueueReadBuffer() call (and not at all if that call failed) ?